### PR TITLE
paperless-ngx: fix build

### PR DESCRIPTION
**Pull request by gador**
###### Description of changes

Build of `paperless-ngx` fails on current master with:

```
―― TestBarcode.test_scan_file_for_separating_barcodes_pillow_transcode_error ―――
[gw27] linux -- Python 3.10.9 /nix/store/al6g1zbk8li6p8mcyp0h60d08jaahf8c-python3-3.10.9/bin/python3.10
self = <documents.tests.test_barcodes.TestBarcode testMethod=test_scan_file_for_separating_barcodes_pillow_transcode_error>

    def test_scan_file_for_separating_barcodes_pillow_transcode_error(self):
        """
        GIVEN:
            - A PDF containing an image which cannot be transcoded to a PIL image
        WHEN:
            - The image tries to be transcoded to a PIL image, but fails
        THEN:
            - The barcode reader is still called
        """

        def _build_device_n_pdf(self, save_path: str):
            # Based on the pikepdf tests
            # https://github.com/pikepdf/pikepdf/blob/abb35ebe17d579d76abe08265e00cf8890a12a95/tests/test_image_access.py
            pdf = pikepdf.new()
            pdf.add_blank_page(page_size=(72, 72))
            imobj = pikepdf.Stream(
                pdf,
                bytes(range(0, 256)),
                BitsPerComponent=8,
                ColorSpace=pikepdf.Array(
                    [
                        pikepdf.Name.DeviceN,
                        pikepdf.Array([pikepdf.Name.Black]),
                        pikepdf.Name.DeviceCMYK,
                        pikepdf.Stream(
                            pdf,
                            b"{0 0 0 4 -1 roll}",  # Colorspace conversion function
                            FunctionType=4,
                            Domain=[0.0, 1.0],
                            Range=[0.0, 1.0, 0.0, 1.0, 0.0, 1.0, 0.0, 1.0],
                        ),
                    ],
                ),
                Width=16,
                Height=16,
                Type=pikepdf.Name.XObject,
                Subtype=pikepdf.Name.Image,
            )
            pim = pikepdf.PdfImage(imobj)
            self.assertEqual(pim.mode, "DeviceN")
            self.assertTrue(pim.is_device_n)

            pdf.pages[0].Contents = pikepdf.Stream(pdf, b"72 0 0 72 0 0 cm /Im0 Do")
            pdf.pages[0].Resources = pikepdf.Dictionary(
                XObject=pikepdf.Dictionary(Im0=imobj),
            )
            pdf.save(save_path)

        with tempfile.NamedTemporaryFile(suffix="pdf") as device_n_pdf:
            # Build an offending file
            _build_device_n_pdf(self, str(device_n_pdf.name))
            with mock.patch("documents.barcodes.barcode_reader") as reader:
                reader.return_value = list()

                _, _ = barcodes.scan_file_for_separating_barcodes(
                    str(device_n_pdf.name),
                )

>               reader.assert_called()

src/documents/tests/test_barcodes.py:280:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <MagicMock name='barcode_reader' id='140737330582848'>

    def assert_called(self):
        """assert that the mock was called at least once
        """
        if self.call_count == 0:
            msg = ("Expected '%s' to have been called." %
                   (self._mock_name or 'mock'))
>           raise AssertionError(msg)
E           AssertionError: Expected 'barcode_reader' to have been called.

/nix/store/al6g1zbk8li6p8mcyp0h60d08jaahf8c-python3-3.10.9/lib/python3.10/unittest/mock.py:898: AssertionError
----------------------------- Captured stderr call -----------------------------
[2023-01-02 08:45:26,070] [WARNING] [paperless.barcodes] Falling back to pdf2image because:
[2023-01-02 08:45:26,074] [WARNING] [paperless.barcodes] Exception during barcode scanning: Unable to get page count. Is poppler installed and in PATH?
------------------------------ Captured log call -------------------------------
DEBUG    paperless.barcodes:barcodes.py:71 Detected mime type: application/pdf
WARNING  paperless.barcodes:barcodes.py:171 Falling back to pdf2image because:
WARNING  paperless.barcodes:barcodes.py:181 Exception during barcode scanning: Unable to get page count. Is poppler installed and in PATH?
```

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by erikarvstedt — APPROVED** [review]
- [x] `pkgs.paperless` succeeds on `x86_64-linux`.
- `nixosTests.paperless` (still?) fails on `x86_64-linux`. This can be handled in another PR.

**Comment by gador** [comment]
@erikarvstedt I'm looking into that ATM

**Comment by gador** [comment]
@erikarvstedt I fixed it in this PR, as paperless-ngx fails without the first fix, anyway. 

The nixOS test timeout was due to a race condition in the startup of the celery worker. The celery scheduler first runs the migration scripts (in `preStart`) and isn't available to the worker yet. The worker fails and after reaching the restart Limit of 5, it will not be restarted. This causes the documents to not be consumed and the curl call to timeout.

The fix was rather easy then. I added a line with `after` in the systemd module which fixes the race condition.

File: nixos/tests/paperless.nix [hunk 1]
```diff
@@ -26,6 +26,9 @@ import ./make-test-python.nix ({ lib, ... }: {
         # Wait until server accepts connections
         machine.wait_until_succeeds("curl -fs localhost:28981")
 
+    with subtest("Task-queue gets ready"):
```
Comment (erikarvstedt) [inline nixos/tests/paperless.nix:29]: If subtest `Add a document via the web interface` requires `paperless-task-queue.service`, better fix this in the service definition instead by adding `systemd.services.paperless-web.after = [ "paperless-task-queue.service" ]`.

Comment (gador) [reply nixos/tests/paperless.nix:29]: The `paperless-web.service` is started by `paperless-scheduler.service`. Independently, `paperless-consumer.service` and `paperless-task-queue.service` are also started by `paperless-scheduler.service`. 
All of them are needed for the consumption by the web-ui. I'm not too sure whether adding `paperless-task-queue.service` to `after` of `paperless-web.service` is needed here.

I guess an argument would be that `paperless-web.service` will fail to start, if the consumption pipeline is not up and running..

File: nixos/modules/services/misc/paperless.nix [hunk 2]
```diff
@@ -258,6 +258,7 @@ in
         PrivateNetwork = false;
       };
       environment = env;
+      after = [ "paperless-scheduler.service" ];
```
[the remark this answers is not in the document]
Comment (gador) [reply nixos/modules/services/misc/paperless.nix:261]: Sure. Should I change the other systemd-units as well? (I just orientated the ordering by the existing units..)

File: nixos/tests/paperless.nix (hunk 1 above)
Comment (erikarvstedt) [reply nixos/tests/paperless.nix:29]: > All of them are needed for the consumption by the web-ui. 

Then it makes sense to encode this in the webinterface service definition.
This ensures that the webinterface is fully functional when the service is active.

File: nixos/modules/services/misc/paperless.nix (hunk 2 above)
[the remark this answers is not in the document]
Comment (erikarvstedt) [reply nixos/modules/services/misc/paperless.nix:261]: Yes, in a separate commit.

File: nixos/tests/paperless.nix (hunk 1 above)
Comment (gador) [reply nixos/tests/paperless.nix:29]: I agree. But isn't that already the case? 
From systemd.services.paperless-web:
`after = [ "paperless-scheduler.service" ];`

Or shall we add a `required` to the web-ui systemd service?

Comment (erikarvstedt) [reply nixos/tests/paperless.nix:29]: I'd also be fine with adding this comment:
```nix
    # Required for consuming documents via the web interface
    with subtest("Task-queue gets ready"):
```

Comment (erikarvstedt) [reply nixos/tests/paperless.nix:29]: `paperless-scheduler` running doesn't imply that `paperless-task-queue` is running.
Yes, the correct way to add this dependency is setting `after` and `required`.

Comment (erikarvstedt) [reply nixos/tests/paperless.nix:29]: But maybe it's better to keep both services separated.
- Simpler code
- The webinterface stays functional on task-queue errors

I'll let you decide.

Comment (gador) [reply nixos/tests/paperless.nix:29]: I just thought about it a bit more. There is an edge case, where the web-ui can function without (any) consumer: 
If just reading and editing the metadata of files, but without consuming of new files.
When we add `requires = [ "paperless-task-queue.service" "paperless-consumer.service" "paperless-scheduler"] ` to `systemd.services.paperless-web` we force the `web`unit to have those available, or fail.

Are we fine with that?

Comment (erikarvstedt) [reply nixos/tests/paperless.nix:29]: For this reason I'd be fine with just https://github.com/NixOS/nixpkgs/pull/208723#discussion_r1060054304.

**Review by erikarvstedt — CHANGES_REQUESTED** [review]
[Fixup](https://github.com/erikarvstedt/nixpkgs/commits/paperless-ngx-add-poppler-ea).
The idea is to put all dependency defs to the top.
Also note the commit msg fix: `reoder` -> `reorder`.

Fetch with:
```bash
git fetch https://github.com/erikarvstedt/nixpkgs paperless-ngx-add-poppler-ea
git show FETCH_HEAD
```

**Comment by gador** [comment]
@erikarvstedt thanks! I applied the fixup.

**Review by erikarvstedt — APPROVED** [review]
- [x] `nixosTests.paperless` succeeds on `x86_64-linux`

Great work, thank you!

**Pull request merged**